File: overlay/widgets/pit_advisor.py
"""Pit engineer — caution and green-flag pit strategy recommendations."""

from __future__ import annotations

import math
from dataclasses import dataclass

from PyQt6.QtCore import QRect, QRectF, QSize, Qt
from PyQt6.QtGui import QFontMetrics, QPainter
from PyQt6.QtWidgets import QSizePolicy, QWidget

from .. import config
from ..pit_strategy import PitRec
from .chrome import col, draw_card, draw_section_header, draw_status_chip, panel_pad
from .fonts import tfont

_SECTION = "pit_advisor"
_REF_H = 100.0
_GAP_TITLE_CHIP = 6.0
_GAP_CHIP_RATIONALE = 8.0
_GAP_RATIONALE_SECONDARY = 6.0
_WRAP_ALIGN = (Qt.AlignmentFlag.AlignLeft | Qt.AlignmentFlag.AlignTop
               | Qt.TextFlag.TextWordWrap)
_PREVIEW = {
    "rec": "pit_next_lap",
    "label": "PIT NEXT LAP",
    "rationale": "Pit next lap to pass #12 \u2014 6.2s ahead, stop costs ~28s",
    "secondary": "Best stop: laps 24\u201326",
    "actionable": True,
    "edit": True,
}


@dataclass
class _PitLayout:
    total_content_h: float
    visible: bool
    show_title: bool
    title_rect: QRectF | None
    chip: QRectF | None
    rationale_rect: QRectF | None
    secondary_rect: QRectF | None
    rationale_font_size: float
    secondary_font_size: float
    rec: str | None
    label: str | None
    rationale: str | None
    secondary: str | None
    active: bool


def _wrap_height(fm: QFontMetrics, text: str, text_w: float) -> float:
    if not text:
        return 0.0
    br = fm.boundingRect(
        QRect(0, 0, int(text_w), 10000), int(_WRAP_ALIGN), text)
    return float(br.height())


def _resolve_fields(d: dict, cfg: dict) -> tuple:
    rec = d.get("rec")
    label = d.get("label")
    rationale = d.get("rationale")
    secondary = d.get("secondary")
    actionable = bool(d.get("actionable"))
    edit = bool(d.get("edit"))

    if not label and edit:
        label = _PREVIEW["label"]
        rationale = _PREVIEW["rationale"]
        secondary = _PREVIEW["secondary"]
        rec = _PREVIEW["rec"]
        actionable = True

    visible = bool(label)
    if visible and cfg.get("show_only_when_actionable", True) and not actionable and not edit:
        visible = False

    active = False
    if rec is not None:
        active = rec not in (PitRec.STAY_OUT.value, PitRec.HOLD.value, PitRec.MARGINAL.value)
        if rec == PitRec.MARGINAL.value:
            active = True

    return rec, label, rationale, secondary, actionable, edit, visible, active


def measure_pit_advisor_layout(width: float, data: dict,
                               cfg: dict | None = None) -> _PitLayout:
    """Compute wrapped text layout and total content height for a panel width."""
    cfg = cfg or config.CFG.get(_SECTION, {})
    d = data or {}
    w = max(1.0, float(width))
    ref_h = _REF_H
    pad = panel_pad(ref_h)
    text_w = max(1.0, w - 1 - 2 * pad)

    y = 0.5 + pad
    show_title = cfg.get("show_title", True)
    title_rect = None
    if show_title:
        hh = max(20.0, ref_h * 0.14)
        title_rect = QRectF(0.5 + pad, y, text_w, hh)
        y += hh + _GAP_TITLE_CHIP

    rec, label, rationale, secondary, _actionable, _edit, visible, active = _resolve_fields(
        d, cfg)

    if not visible:
        return _PitLayout(
            total_content_h=0.0,
            visible=False,
            show_title=False,
            title_rect=None,
            chip=None,
            rationale_rect=None,
            secondary_rect=None,
            rationale_font_size=11.0,
            secondary_font_size=10.0,
            rec=rec,
            label=label,
            rationale=rationale,
            secondary=secondary,
            active=active,
        )

    body_h = max((ref_h - 1) - pad - y, 40.0)
    chip_h = max(24.0, body_h * 0.38)
    chip = QRectF(0.5 + pad, y, text_w, chip_h)
    y += chip_h + _GAP_CHIP_RATIONALE

    rationale_font_size = max(11.0, chip_h * 0.38)
    secondary_font_size = max(10.0, chip_h * 0.34)
    rationale_rect = None
    secondary_rect = None

    if rationale:
        fm = QFontMetrics(tfont(rationale_font_size, bold=False))
        rh = _wrap_height(fm, str(rationale), text_w)
        rationale_rect = QRectF(0.5 + pad, y, text_w, rh)
        y += rh
        if secondary:
            y += _GAP_RATIONALE_SECONDARY

    if secondary:
        fm = QFontMetrics(tfont(secondary_font_size, bold=False))
        sh = _wrap_height(fm, str(secondary), text_w)
        secondary_rect = QRectF(0.5 + pad, y, text_w, sh)
        y += sh

    return _PitLayout(
        total_content_h=max(72.0, y + pad),
        visible=True,
        show_title=show_title,
        title_rect=title_rect,
        chip=chip,
        rationale_rect=rationale_rect,
        secondary_rect=secondary_rect,
        rationale_font_size=rationale_font_size,
        secondary_font_size=secondary_font_size,
        rec=rec,
        label=label,
        rationale=rationale,
        secondary=secondary,
        active=active,
    )


class PitAdvisorWidget(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.data: dict = {}
        self.setMinimumSize(180, 72)
        self.setSizePolicy(QSizePolicy.Policy.Expanding,
                           QSizePolicy.Policy.Expanding)

    def set_data(self, data: dict) -> None:
        data = data or {}
        if data != self.data:
            self.data = data
        self.update()
        self._sync_panel_height()

    def resizeEvent(self, event) -> None:  # noqa: N802
        super().resizeEvent(event)
        self._sync_panel_height()

    def sizeHint(self):  # noqa: N802
        return self._measured_size()

    def minimumSizeHint(self):  # noqa: N802
        return self._measured_size()

    def _measured_size(self) -> QSize:
        w = max(self.width(), self.minimumWidth())
        layout = measure_pit_advisor_layout(w, self.data)
        return QSize(w, int(math.ceil(layout.total_content_h)))

    def _sync_panel_height(self) -> None:
        layout = measure_pit_advisor_layout(self.width(), self.data)
        self.updateGeometry()
        win = self.window()
        from ..panel import PanelWindow
        if isinstance(win, PanelWindow):
            win.fit_content_height(int(math.ceil(layout.total_content_h)))

    def paintEvent(self, event) -> None:  # noqa: N802
        config.use_section(_SECTION)
        w, h = float(self.width()), float(self.height())
        cfg = config.CFG.get(_SECTION, {})
        layout = measure_pit_advisor_layout(w, self.data, cfg)

        if not layout.visible or not layout.chip or not layout.label:
            return

        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)
        card, radius = draw_card(p, w, h, _SECTION)

        if layout.show_title and layout.title_rect is not None:
            draw_section_header(
                p, layout.title_rect,
                str(cfg.get("title", "PIT ENGINEER")),
                _SECTION, radius_top=radius)

        draw_status_chip(p, layout.chip, str(layout.label), _SECTION,
                         active=layout.active)

        if layout.rationale_rect is not None and layout.rationale:
            p.setFont(tfont(layout.rationale_font_size, bold=False))
            p.setPen(col("text", _SECTION))
            p.drawText(layout.rationale_rect, _WRAP_ALIGN, str(layout.rationale))

        if layout.secondary_rect is not None and layout.secondary:
            p.setFont(tfont(layout.secondary_font_size, bold=False))
            p.setPen(col("muted", _SECTION))
            p.drawText(layout.secondary_rect, _WRAP_ALIGN, str(layout.secondary))

File: overlay-rs/crates/gridglance-overlay/src/widgets/pit_advisor.rs
//! Pit engineer panel (Python `pit_advisor.py` parity).

use super::WidgetCtx;
use crate::chrome::{
    color_with_alpha, full_rect, is_elegant, label, panel_card, panel_content_pad, panel_title,
};
use egui::{Align2, Color32, CornerRadius, Pos2, Rect, Ui, Vec2};

const SECTION: &str = "pit_advisor";

pub fn paint(ui: &mut Ui, ctx: &mut WidgetCtx<'_>) {
    let advice = ctx.frame.pit_advice.clone().unwrap_or_default();
    let only_actionable = ctx.cfg.bool_key(SECTION, "show_only_when_actionable", true);
    if only_actionable && !advice.actionable && !ctx.edit_mode {
        let _ = full_rect(ui);
        return;
    }

    let rect = full_rect(ui);
    let (card, radius) = panel_card(ui, ctx.cfg, SECTION, rect);
    let pad = panel_content_pad(ctx.cfg, SECTION, card.height());
    let elegant = is_elegant(ctx.cfg, SECTION);
    let mut y = card.top() + pad;
    let text_w = card.width() - 2.0 * pad;

    if elegant {
        if ctx.cfg.bool_key(SECTION, "show_title", true) {
            label(
                ui,
                Pos2::new(card.left() + pad, y + 6.0),
                Align2::LEFT_CENTER,
                &ctx.cfg.str_key(SECTION, "title", "PIT ENGINEER"),
                10.0,
                color_with_alpha(ctx.cfg.color(SECTION, "muted", "#8b93a1"), 200),
                false,
            );
            y += 14.0;
        }
    } else {
        y = panel_title(ui, ctx.cfg, SECTION, card, radius, y, pad, "PIT ENGINEER");
    }

    let label_txt = if advice.label.is_empty() && ctx.edit_mode {
        "PIT NEXT LAP".into()
    } else {
        advice.label.clone()
    };
    let rationale = if advice.rationale.is_empty() && ctx.edit_mode {
        "Pit next lap to pass #12 — 6.2s ahead, stop costs ~28s".into()
    } else {
        advice.rationale.clone()
    };
    let rec = if advice.rec.is_empty() {
        "pit_next_lap".into()
    } else {
        advice.rec.clone()
    };
    let active = !matches!(rec.as_str(), "stay_out" | "hold");

    let chip_h = if elegant {
        24.0
    } else {
        (card.height() * 0.18).max(22.0)
    };
    let chip = Rect::from_min_size(Pos2::new(card.left() + pad, y), Vec2::new(text_w, chip_h));
    let chip_bg = if active {
        ctx.cfg.color(SECTION, "chip_active", "#ff9416")
    } else {
        ctx.cfg.color(SECTION, "chip_idle", "#333a42")
    };
    ui.painter().rect_filled(
        chip,
        CornerRadius::same(if elegant { 8 } else { 6 }),
        chip_bg,
    );
    label(
        ui,
        chip.center(),
        Align2::CENTER_CENTER,
        &label_txt,
        (chip_h * 0.42).clamp(11.0, 16.0),
        Color32::WHITE,
        true,
    );
    y += chip_h + if elegant { 6.0 } else { 8.0 };

    label(
        ui,
        Pos2::new(card.left() + pad, y),
        Align2::LEFT_TOP,
        &rationale,
        if elegant { 11.0 } else { 13.0 },
        if elegant {
            color_with_alpha(ctx.cfg.color(SECTION, "text", "#f4f6f8"), 210)
        } else {
            ctx.cfg.color(SECTION, "text", "#f4f6f8")
        },
        false,
    );
    y += 28.0;

    if let Some(sec) = advice
        .secondary
        .clone()
        .or_else(|| ctx.edit_mode.then(|| "Best stop: laps 24–26".into()))
    {
        label(
            ui,
            Pos2::new(card.left() + pad, y),
            Align2::LEFT_TOP,
            &sec,
            10.0,
            color_with_alpha(ctx.cfg.color(SECTION, "muted", "#8b93a1"), 200),
            false,
        );
    }
}
